File: mobile/lib/features/integrations/mappers/activity_type_mapper.dart
import 'package:health/health.dart';

abstract class ActivityTypes {
  static const running  = 'Running';
  static const cycling  = 'Cycling';
  static const swimming = 'Swimming';
  static const hiking   = 'Hiking';
  static const climbing = 'Climbing';
  static const yoga     = 'Yoga';
  static const gym      = 'Gym';
}

abstract class IntegrationProviders {
  static const healthKit     = 'HealthKit';
  static const healthConnect = 'HealthConnect';
  static const strava        = 'Strava';
  static const garmin        = 'Garmin';
}

class ActivityTypeMapper {
  static String fromHealthConnect(HealthWorkoutActivityType type) {
    return switch (type) {
      HealthWorkoutActivityType.RUNNING              => ActivityTypes.running,
      HealthWorkoutActivityType.RUNNING_TREADMILL    => ActivityTypes.running,
      HealthWorkoutActivityType.WALKING              => ActivityTypes.hiking,
      HealthWorkoutActivityType.BIKING               => ActivityTypes.cycling,
      HealthWorkoutActivityType.SWIMMING             => ActivityTypes.swimming,
      HealthWorkoutActivityType.HIKING               => ActivityTypes.hiking,
      HealthWorkoutActivityType.CLIMBING             => ActivityTypes.climbing,
      HealthWorkoutActivityType.ROCK_CLIMBING        => ActivityTypes.climbing,
      HealthWorkoutActivityType.YOGA                 => ActivityTypes.yoga,
      HealthWorkoutActivityType.PILATES              => ActivityTypes.yoga,
      HealthWorkoutActivityType.TRADITIONAL_STRENGTH_TRAINING => ActivityTypes.gym,
      HealthWorkoutActivityType.FUNCTIONAL_STRENGTH_TRAINING  => ActivityTypes.gym,
      HealthWorkoutActivityType.HIGH_INTENSITY_INTERVAL_TRAINING => ActivityTypes.gym,
      HealthWorkoutActivityType.CROSS_TRAINING       => ActivityTypes.gym,
      HealthWorkoutActivityType.ELLIPTICAL           => ActivityTypes.gym,
      HealthWorkoutActivityType.ROWING               => ActivityTypes.swimming,
      HealthWorkoutActivityType.ROWING_MACHINE       => ActivityTypes.gym,
      HealthWorkoutActivityType.SURFING              => ActivityTypes.swimming,
      _                                              => ActivityTypes.gym,
    };
  }
}

File: backend/src/modules/LifeLevel.Modules.Integrations/Application/Mappers/ActivityTypeMapper.cs
namespace LifeLevel.Modules.Integrations.Application.Mappers;

public static class ActivityTypeMapper
{
    public static class Types
    {
        public const string Running  = "Running";
        public const string Cycling  = "Cycling";
        public const string Swimming = "Swimming";
        public const string Hiking   = "Hiking";
        public const string Climbing = "Climbing";
        public const string Yoga     = "Yoga";
        public const string Gym      = "Gym";
    }

    /// <summary>Maps a Strava sport_type string to an internal activity type.</summary>
    public static string FromStrava(string sportType) => sportType switch
    {
        "Run" or "TrailRun" or "VirtualRun"                                           => Types.Running,
        "Ride" or "MountainBikeRide" or "GravelRide" or "VirtualRide" or "EBikeRide" => Types.Cycling,
        "Swim" or "OpenWaterSwim"                                                      => Types.Swimming,
        "Hike" or "Walk"                                                               => Types.Hiking,
        "RockClimbing" or "IceClimbing"                                                => Types.Climbing,
        "Yoga" or "Pilates"                                                            => Types.Yoga,
        _                                                                               => Types.Gym,
    };

    /// <summary>Maps a Garmin Connect activity type key to an internal activity type.</summary>
    public static string FromGarmin(string activityType) => activityType?.ToLowerInvariant() switch
    {
        "running" or "trail_running" or "treadmill_running" or "track_running"              => Types.Running,
        "cycling" or "road_cycling" or "mountain_biking" or "indoor_cycling" or "gravel_cycling" => Types.Cycling,
        "swimming" or "lap_swimming" or "open_water_swimming"                               => Types.Swimming,
        "hiking" or "walking"                                                               => Types.Hiking,
        "rock_climbing" or "bouldering"                                                     => Types.Climbing,
        "yoga" or "pilates"                                                                 => Types.Yoga,
        _                                                                                   => Types.Gym,
    };
}
